---
author:
  name: Goran Soderstrom
  picture: 111361
body: "I just switch to a brand new computer and set up everything but can\u2019t
  get TTX to work.\r\n\r\nI see this when i type ttx in the terminal so I think it\u2019s
  installed right?\r\n\r\n---------------\r\n\r\n TTX 2.3 -- From OpenType To XML
  And Back\r\n\r\n    If an input file is a TrueType or OpenType font file, it will
  be\r\n       dumped to an TTX file (an XML-based text format).\r\n    If an input
  file is a TTX file, it will be compiled to a TrueType\r\n       or OpenType font
  file.\r\n\r\netc .... \r\n\r\n---------------\r\n\r\nBut when actually trying to
  work on a font file I get this:\r\n\r\n---------------\r\n\r\n  File \"/usr/local/bin/ttx\",
  line 11, in <module>\r\n    ttx.main(sys.argv[1:])\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
  line 295, in main\r\n    process(jobs, options)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
  line 280, in process\r\n    action(input, output, options)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
  line 170, in ttDump\r\n    fontNumber=options.fontNumber)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py\",
  line 145, in __init__\r\n    if macUtils.getSFNTResIndices(file):\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/macUtils.py\",
  line 28, in getSFNTResIndices\r\n    resref = MyOpenResFile(path)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/macUtils.py\",
  line 18, in MyOpenResFile\r\n    resref = Res.FSpOpenResFile(path, mode)\r\nAttributeError:
  'module' object has no attribute 'FSpOpenResFile'\r\n\r\n---------------\r\n\r\nOn
  Lion with Python 2.7.\r\n\r\nAny ideas what could be wrong? Would really appreciate
  it."
comments:
- author:
    name: Michel Boyer
    picture: 112585
  body: "I have not installed ttx2.3 on Lion. I am using ttx2.2 from AFDKO with no
    problem. Here are the differences between 2.2 and 2.3 according to the file changes.txt\r\n<code>\r\nTTX/FontTools
    Version 2.3</code>\r\n<code>\r\n- TrueType Collection (TTC) support\r\n- Python
    2.6 support\r\n- Update Unicode data to 5.2.0\r\n- Couple of bug fixes\r\n</code>\r\nMaybe
    I should add that the AFDKO ttx comes with its own python, it does not rely on
    some pre-installed python."
  created: '2012-01-13 17:40:36'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "This is the TTX that comes with Fonttools in the Robofab package.\r\n\r\nSo.
    it doesn't support Python 2.7 which came with the computer? Is that what you're
    meaning?\r\n\r\n\r\n"
  created: '2012-01-13 17:46:52'
- author:
    name: Michel Boyer
    picture: 112585
  body: "Well, on  my mac with Lion, there are three versions of python. If I type
    python2.5 I get python2.5, if I type python2.6, I get python 2.6. Here is a trace
    of execution:\r\n<code>\r\n% ls /usr/bin/python*\r\n/usr/bin/python\t\t\t/usr/bin/python2.5\t\t/usr/bin/python2.6\t\t/usr/bin/python2.7\t\t/usr/bin/pythonw\t\t/usr/bin/pythonw2.6\r\n/usr/bin/python-config\t\t/usr/bin/python2.5-config\t/usr/bin/python2.6-config\t/usr/bin/python2.7-config\t/usr/bin/pythonw2.5\t\t/usr/bin/pythonw2.7\r\n</code>\r\nYou
    are not limited to python2.7 and to use the python fontforge module with a slightly
    modified mactools installation of fontforge, I have to use python2.6 (works on
    OS X 10.5 and OS X 10.7; have'nt tried yet on OS X 10.6). I know nothing of fontlab
    (don't want to pay just to answer questions on typophile).\r\n"
  created: '2012-01-13 17:55:15'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "Sorry, but I don\u2018t understand what to do. I just want TTX to work again.
    In the terminal. Nothing to do with FontLab."
  created: '2012-01-13 18:12:45'
- author:
    name: "T\xE9 Rowan"
    picture: 121227
  body: "First, then, do the <code>ls /usr/bin/python*</code> thing. Also, if your
    machine has <code>/usr/local/</code>, try <code>ls /usr/local/bin/python*</code>
    in case there is a Python interpreter lurking there.\r\n\r\nAssuming you found,
    say, <code>/usr/bin/python2.5</code>: Say <code>python2.5 ttx foo.ttf</code> to
    check if it runs to completion. If <code>/usr/bin/python2.6</code> exists, try
    again with that.\r\n\r\nThen, open the <code>ttx</code> script in a text editor
    and change the <code>#! /usr/bin/env python</code> (the sha-bang line) at the
    top to reflect the latest version that works. Save it, and you should be back
    to NERV's motto."
  created: '2012-01-13 18:59:58'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "I\u2019m with you through here:\r\n\r\nls /usr/bin/python*\r\n\r\nand gets:\r\n\r\n/usr/bin/python\t\t\t/usr/bin/python2.6\t\t/usr/bin/pythonw\r\n/usr/bin/python-config\t\t/usr/bin/python2.6-config\t/usr/bin/pythonw2.5\r\n/usr/bin/python2.5\t\t/usr/bin/python2.7\t\t/usr/bin/pythonw2.6\r\n/usr/bin/python2.5-config\t/usr/bin/python2.7-config\t/usr/bin/pythonw2.7\r\n\r\nBut,
    after that when I type:\r\n\r\npython2.5 ttx foo.ttf\r\n\r\nthis happens:\r\n\r\npython2.5:
    can't open file 'ttx': [Errno 2] No such file or directory\r\n\r\nSame with 2.6."
  created: '2012-01-13 19:40:27'
- author:
    name: Jens Kutilek
    picture: 113138
  body: "Hi G\xF6ran,\r\n\r\nyou need to edit one line in the fontTools source:\r\n\r\nhttps://github.com/mcolyer/fonttools/commit/e732bd3ba63c51df9aed903eb2147fa1af1bfdc2\r\n\r\nAfter
    that you need to run <code>sudo ./setup.py install</code> again in the fontTools
    folder."
  created: '2012-01-13 20:57:10'
- author:
    name: Michel Boyer
    picture: 112585
  body: "Jens\r\n\r\nDoes that solve all problems? I see that the deprecated calls
    FSpOpenResFile and FSpCreateResFile are also used in the files \r\n<code>\r\n./Lib/fontTools/fondLib.py
    ./Lib/fontTools/nfntLib.py ./Lib/fontTools/t1Lib.py ./Mac/TTX.py\r\n</code>\r\n\r\nMichel\r\n"
  created: '2012-01-13 21:09:41'
- author:
    name: Jens Kutilek
    picture: 113138
  body: "You're probably right, somebody should test fontTools thoroughly for stuff
    like this. As I understand it, it's a 64 bit issue, so the other way to avoid
    it is by switching Python to 32 bit by default:\r\n\r\n <code>defaults write com.apple.versioner.python
    Prefer-32-Bit -bool yes</code>"
  created: '2012-01-13 21:14:10'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "Jens, do you know everything? :)\r\n\r\nIt worked! Thanks so much. Again."
  created: '2012-01-14 09:11:02'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "Ok well. It worked to *create* the TTX but when going back to the TTF I get
    this:\r\n\r\nParsing 'GlyphOrder' table...\r\nParsing 'head' table...\r\nParsing
    'hhea' table...\r\nParsing 'maxp' table...\r\nParsing 'OS/2' table...\r\nParsing
    'hmtx' table...\r\nParsing 'cmap' table...\r\nParsing 'fpgm' table...\r\nParsing
    'prep' table...\r\nParsing 'cvt ' table...\r\nParsing 'loca' table...\r\nParsing
    'glyf' table...\r\nParsing 'kern' table...\r\nParsing 'name' table...\r\nParsing
    'post' table...\r\nParsing 'gasp' table...\r\nParsing 'GPOS' table...\r\nParsing
    'GSUB' table...\r\nTraceback (most recent call last):\r\n  File \"/usr/local/bin/ttx\",
    line 11, in <module>\r\n    ttx.main(sys.argv[1:])\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
    line 295, in main\r\n    process(jobs, options)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
    line 280, in process\r\n    action(input, output, options)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py\",
    line 186, in ttCompile\r\n    ttf.save(output)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py\",
    line 201, in save\r\n    self._writeTable(tag, writer, done)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py\",
    line 597, in _writeTable\r\n    tabledata = self.getTableData(tag)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py\",
    line 609, in getTableData\r\n    return self.tables[tag].compile(self)\r\n  File
    \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otBase.py\",
    line 73, in compile\r\n    self.table.compile(writer, font)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otBase.py\",
    line 616, in compile\r\n    conv.write(writer, font, tableStack, value)\r\n  File
    \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otConverters.py\",
    line 185, in write\r\n    value.compile(subWriter, font, tableStack)\r\n  File
    \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otBase.py\",
    line 604, in compile\r\n    conv.write(writer, font, tableStack, value[i], i)\r\n
    \ File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otConverters.py\",
    line 185, in write\r\n    value.compile(subWriter, font, tableStack)\r\n  File
    \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otBase.py\",
    line 604, in compile\r\n    conv.write(writer, font, tableStack, value[i], i)\r\n
    \ File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otConverters.py\",
    line 185, in write\r\n    value.compile(subWriter, font, tableStack)\r\n  File
    \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otBase.py\",
    line 590, in compile\r\n    table = self.preWrite(font)\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/otTables.py\",
    line 186, in preWrite\r\n    gidItems = [(getGlyphID(item[0]), getGlyphID(item[1]))
    for item in items]\r\n  File \"/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py\",
    line 548, in getGlyphID\r\n    raise KeyError, glyphName\r\nKeyError: 'glyph65535'\r\n\r\nAnd
    it only seem to be happening on some fonts I now realized. Must be something wrong
    with the font perhaps? It\u2019s a TTF generated from Fontlab 5.1.\r\n\r\n"
  created: '2012-01-14 09:27:59'
- author:
    name: Jens Kutilek
    picture: 113138
  body: "<em>Jens, do you know everything? :)</em>\r\n\r\nI'm working on it ;)\r\n\r\n<em>KeyError:
    'glyph65535'</em>\r\n\r\n65535 or 2\xB9\u2076 \u2212 1 is the maximum glyph index
    in OpenType fonts.\r\n\r\nThis error is usually because of a glyph being referenced
    in an OT feature but missing from the font. The error is in the GSUB table here,
    so I'd check the classes and substitution features for missing glyphs.\r\n"
  created: '2012-01-14 15:50:17'
- author:
    name: Michel Boyer
    picture: 112585
  body: "That is due to the fact that the python that comes installed on the mac is
    ucs-2 and not ucs-4. Here is a trace of execution on my mac (python2.5, python2.6
    and python2.7 in /usr/bin do the same):\r\n<code>\r\n501 % /usr/bin/python -c
    \"import sys; print sys.maxunicode;\"\r\n65535\r\n502 % /opt/local/bin/python
    -c \"import sys; print sys.maxunicode;\"\r\n1114111\r\n503 % \r\n</code>\r\nI
    installed a ucs-4 python (2.6) with macports to use with fontforge to handle glyphs
    above 0xFFFF."
  created: '2012-01-14 16:26:49'
- author:
    name: Goran Soderstrom
    picture: 111361
  body: "Jens! Hero! You were right again!\r\n\r\nFor some reason the font could be
    generated even though a glyph was represented in the code (but not in the font).\r\n\r\nI\u2019m
    happy.\r\n\r\nBeer\u2019s on me."
  created: '2012-01-14 17:06:14'
- author:
    name: Michel Boyer
    picture: 112585
  body: "&gt;<cite>&nbsp;&nbsp; 65535 or 2\xB9\u2076 \u2212 1 is the maximum glyph
    index in OpenType fonts</cite>\r\nSorry, I had misread your comment. GlyphID is
    indeed 16 bits."
  created: '2012-01-15 12:56:28'
date: '2012-01-13 17:13:41'
node_type: forum
title: "On Lion and can\u2019t get TTX to work anymore."

---
